### PR TITLE
load binary shaders (spv)

### PR DESCRIPTION
**Pull request by barnoy1**
Allow openscenegraph to load binary shader  (*.spv)
This PR also include a sample (osgbinaryshader) in which we load a binary shader or source code (text file).
The sample contains two light source and a model (cube) using a simple blinn-phong model shading implementation
 
image

### CODE REVIEW

**Comment by LaurensVoerman** [comment]
Looking trough the diff I see a few problems:

1> this will only work on windows:

examples/osgbinshader/osgbinshader.cpp

#include "C:/Program Files/NVIDIA Corporation/NvToolsExt/include/nvToolsExt.h"

2> whats this GLEW dependency?

GLEW is added as a required dependency in a way that only works on
windows; while (as far as I can see) it's not used at all

3> examples/osgbinshader/osgbinshader.cpp refers to data files with
"../../../examples/osgbinshader/" path

cube.obj is not submitted at all,

And if there is no cube.obj in that path the user gets the very
unhelpful "Unable to load model from command line."


Regards, Laurens.

**Review by glebm — CHANGES_REQUESTED** [review]
Thanks for your contribution! I've left a few comments to save the maintainers of this repo some effort.

**Comment by barnoy1** [comment]
Hi @LaurensVoerman and @glebm ,
I would like to address the issues you mentioned in previous posts:

First I would like to reply to @LaurensVoerman post
https://github.com/openscenegraph/OpenSceneGraph/pull/1054#issuecomment-800370034
1 > I remove #include "C:/Program Files/NVIDIA Corporation/NvToolsExt/include/nvToolsExt.h"

3 > I rework the samples so now the path to the shader file (binary or source) are provided as inputs.

2 > GLEW here is need because *GL_SHADER_BINARY_FORMAT_SPIR_V_ARB* is an extension that only found when glew is included. It is need for opengl to compile the shader. The reason I put it in Shader.cpp is to be as close as possible to the extension and avoid to pollute the entire project with unnecessary include.

During the last commits I took into consideration  @glebm's comments:
1. GLEW was added via the CMake's built-in FindGLEW  I use FIND_GLEW as macro to hide *GL_SHADER_BINARY_FORMAT_SPIR_V_ARB* allowing the project to compile and work as before.
2.  #include <GL\glew.h> was changed into #include <GL/glew.h>
3.  Shader.cpp was re submitted so now it is only include the suggested changes.

File: src/osg/Shader.cpp [hunk 1]
```diff
@@ -34,73 +36,74 @@
...
-// It does only one pass, i.e. new matches created in a position before the current match are not replaced
-void replaceAll(std::string& str, const std::string& from, const std::string& to)
-{
-    std::string::size_type pos = 0;
-    while ((pos = str.find(from, pos)) != std::string::npos)
+    // Replaces all occurrences of "from" with the contents of "to"
+    // It does only one pass, i.e. new matches created in a position before the current match are not replaced
+    void replaceAll(std::string& str, const std::string& from, const std::string& to)
     {
-        str.replace(pos, from.length(), to);
```
[the remark this answers is not in the document]
Comment (ronbarnoy) [reply src/osg/Shader.cpp]: hader.cpp was re submitted so now it is only include the suggested changes.

File: CMakeLists.txt [hunk 2]
```diff
@@ -165,6 +165,31 @@ IF(ANDROID)
     INCLUDE(OsgAndroidMacroUtils)
 ENDIF()
 
+
+#############################################################
+#                   Adding GLEW                             #
+#############################################################
+
+SET (GLEW_ROOT_DIR ${PROJECT_SOURCE_DIR}/external/glew-2.2.0)
```
[the remark this answers is not in the document]
Comment (ronbarnoy) [reply CMakeLists.txt]: 1. As I explained [here](https://github.com/openscenegraph/OpenSceneGraph/pull/1054#issuecomment-801916905), GLEW dependency is necessary for GL_SHADER_BINARY_FORMAT_SPIR_V_ARB option.
2. GLEW was added via the CMake's built-in FindGLEW. I use FIND_GLEW as macro to hide GL_SHADER_BINARY_FORMAT_SPIR_V_ARB allowing the project to compile and work as before.

File: src/osg/Shader.cpp [hunk 3]
```diff
@@ -19,6 +19,8 @@
  *         Holger Helmich 2010-10-21
 */
 
+#include <GL\glew.h>
```
[the remark this answers is not in the document]
Comment (ronbarnoy) [reply src/osg/Shader.cpp]: #include <GL\glew.h> was changed into #include <GL/glew.h>

**Comment by robertosfield** [comment]
Just done a first pass through review.  Lots of changes.  

The OSG has gone 2 decades without needing something like GLEW as it has it's own scheme for extension checking that handles multiple contexts correctly.  We are now about add a dependency for a niche feature that can easily be properly implemented without GLEW.

This PR will not be merged with a GLEW dependency.

**Comment by robertosfield** [comment]
I'm closing this PR as none of the problems raised with it have not been addressed.

**Pull request closed without merging**